### ISSUE TITLE
Migrator should use pub outdated to advise users

### ISSUE THREAD

**Issue by srawlins** [dart-lang/sdk#41769]
There is room for design or changes to this feature, but here is the basic idea:

When a user goes to migrate a package (call it "my_pkg"), by just `cd`-ing into their repo and running the migration tool, we should do some checks up front, regarding their dependencies. Let's say my_pkg has a dependency on my_dep, and my_dep has the following releases:

* my_dep 1.0.0 is _not_ opted into Null Safety (SDK constraint is something like ">=2.3.0 <3.0.0"
* my_dep 1.1.0 _is_ opted into Null Safety (SDK constraint is something like ">=2.9.0" or ">=2.10.0"

my_pkg currently specifies "my_dep: ^1.0" in pubspec.yaml and the user's checkout has a pubspec.lock with "my_dep: 1.0.0". In this situation, the migration tool should tell the user "Hold up! There is a newer version of my_dep which you should fetch before running this tool!" The mechanism for determining this is the new `pub outdated` command available in Dart 2.8.0. Documentation is being written at https://github.com/dart-lang/site-www/pull/2332. I think we need the `--null-safety` flag, which is being authored at https://github.com/dart-lang/pub/pull/2461.

If there are outdated dependencies, I think the migrator should report this, and exit. But maybe prompt to continue. Or exit and offer a --force flag to allow migrating when there are newer dependency releases available.

**Comment by srawlins** [comment]
dart-lang/pub#2461 has landed, so I believe this is unblocked.

**Comment by srawlins** [comment]
This version of pub has not been revved into the SDK, but I can work on the integration locally until it is.

**Comment by jonasfj** [comment]
Should be landing in [+148532](https://dart-review.googlesource.com/c/sdk/+/148532).

This sounds pretty awesome. Quickly noting that even if the first of `my_dep` with null-safety support is `2.0.0` it's probably a good idea to suggest user migrates to a newer version of the dependency.

**Comment by srawlins** [comment]
Yeah I like that idea. I think we'll give a few tips like this in our README or something:

> If you are using my_dep 1.0.0, but my_dep 2.0.0 is the first version with null-safety, then before migrating _this_ package to null-safety, first bump my_dep to 2.0.0 and commit that change, handling anything else required for my_dep 2.0.0.

**Comment by natebosch** [comment]
Reopening rather than filing a new issue.

This feature is not usable with our current migration plan. Some packages which are expected to be in the transitive `dev_dependencies` for _every_ package won't be migrated before this all goes stable. Packages like `analyzer` likely won't be analyze, but are present in the transitive dev dependencies for any package which uses `package:test`.

**Comment by stereotype441** [comment]
> Reopening rather than filing a new issue.
> 
> This feature is not usable with our current migration plan. Some packages which are expected to be in the transitive `dev_dependencies` for _every_ package won't be migrated before this all goes stable. Packages like `analyzer` won't be migrated before the null safety stable release, but are present in the transitive dev dependencies for any package which uses `package:test`.

@sigurdm @jonasfj is it possible to modify the behavior of `pub outdated` to address this problem?

**Comment by natebosch** [comment]
I suspect that the only feasible solution here is to ignore `dev_dependencies`, including transitive, for this. I don't know whether pub already has a notion of transitive deps vs transitive dev deps. There is not a realistic way for it know know which `package:test` dependencies are imported by test code, and which are used only by the runner.

**Comment by sigurdm** [comment]
```
> pub outdated --help
Analyze your dependencies to find which ones can be upgraded.

Usage: pub outdated [options]
-h, --help                         Print this usage information.
    --[no-]color                   Whether to color the output.
                                   Defaults to color when connected to a terminal, and no-color otherwise.
    --[no-]dependency-overrides    Show resolutions with `dependency_overrides`.
                                   (defaults to on)
    --[no-]dev-dependencies        Take dev dependencies into account.
                                   (defaults to on)
    --json                         Output the results using a json format.
    --mode=<PROPERTY>              Highlight versions with PROPERTY.
                                   Only packages currently missing that PROPERTY will be included unless --show-all.
                                   [outdated (default), null-safety]
    --[no-]prereleases             Include prereleases in latest version.
    --[no-]show-all                Include dependencies that are already fullfilling --mode.

Run "pub help" to see global options.
See https://dart.dev/tools/pub/cmd/pub-outdated for detailed documentation.
```

if you do --no-dev-dependencies it will "strip" dev_dependencies from the pubspec before doing resolutions, this also removes packages that are transitively required only by the dev_dependencies. Dependencies that are shared between dev and non-dev-dependencies are maintained.

**Comment by jonasfj** [comment]
@sigurdm, we might need to fix https://github.com/dart-lang/pub/issues/2552, so that `--no-dev-dependencies` works with `--json`.

**Comment by sigurdm** [comment]
Yes, forgot about that - though I think the workaround you mention on that issue should work just fine. Packages with no upgradable or resolvable versions are not out of date.

**Comment by srawlins** [comment]
I think we will use `--no-dev-dependencies` in the migration tool.

**Comment by franklinyow** [comment]
@srawlins @leafpetersen 
Is this still a blocker for anything?

**Comment by franklinyow** [comment]
ping again @srawlins @leafpetersen. 
What is the status on this?  Thanks!

**Comment by leafpetersen** [comment]
I believe this is implemented, but temporarily disabled for TP2.  @devoncarew @srawlins is that right?

**Comment by franklinyow** [comment]
Can I get a confirmation?  Like to declare we have zero blocker ASAP,  thanks!

**Comment by stereotype441** [comment]
I can confirm that this is implemented but temporarily disabled.  I have a CL to re-enable it here: https://dart-review.googlesource.com/c/sdk/+/167380 (note that this CL depends on the flag flip CL)
